### PR TITLE
Adding AMIUpdater

### PR DESCRIPTION
**Pull request by andrew-glenn**
## Overview
This PR Adds AMIUpdater. A README is forthcoming in a future PR.

### CODE REVIEW

**Comment by avattathil** [comment]
This PR is superseded by https://github.com/aws-quickstart/taskcat/pull/274

**Comment by jaymccon** [comment]
this looks amazing @andrew-glenn, but pretty large, going to need some time for review

**Comment by jaymccon** [comment]
what I did to get to the test failures:

```
git clone git@github.com:andrew-glenn/taskcat.git
cd taskcat
git checkout develop
python setup.py test
```

**Comment by avattathil** [comment]
I would consider modifying the unittests to not be dependent on valid credentials.  
Add a test to this sections [ functional tests](https://github.com/aws-quickstart/taskcat/blob/master/.travis.yml#L65-L68)( the functional tests will only run as part of PR validation from with in travis env). This change allow community developer to run pytest with no dependency on cloud access.  
With that said the PR look awesome! Nice work!

**Comment by santiagocardenas** [comment]
Rest in peace the Alchemist. Long live the AMIcorn!

File: bin/amiupdater [hunk 1]
```diff
@@ -1 +1,38 @@
-# Placeholder for future development.
+#!/usr/bin/env python
```
Comment (jaymccon) [inline bin/amiupdater]: can as much logic as possible be removed from this and added to the AMIUpdate class as static methods ? Will make it easier to test those components and leave a smaller coverage gap if we don't have tests for this file. Will also make integration into v9 easier.

File: taskcat/amiupdater.py [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+        self.region = region
+
+    def __lt__(self, other):
+        # See Codenames.parse_api_results for notes on why this is here.
+        if self.custom_comparisons:
+            return self._creation_date < other._creation_date
+        else:
+            return object.__lt__(self, other)
+
+    def __gt__(self, other):
```
Comment (jaymccon) [inline taskcat/amiupdater.py]: can we get a test for this ?

File: taskcat/amiupdater.py [hunk 3]
```diff
@@ -0,0 +1,369 @@
+import collections
```
Comment (jaymccon) [inline taskcat/amiupdater.py]: pycharm picks up 6 warnings and 29 weak warnings for this file, can we get as many as possible addressed, alot are just PEP8, but some that look like they may manifest as bugs and are not in the tested codepath.

File: taskcat/amiupdater.py [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+    def objects(cls):
+        return [z for z in cls._x.values() if z.filters]
+
+    def __init__(self, cn, region, filters=None, *args, **kwargs):
+        self.region = region
+        self._results = None
+        self._filters = filters
+        super(RegionalCodename, self).__init__(cn, region, *args, **kwargs)
+
+class TemplateClass(object):
```
Comment (jaymccon) [inline taskcat/amiupdater.py]: looks like only regions method needs to be a classmethod, others can all be static.

File: taskcat/amiupdater.py [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+        stripped_tfdata = tfdata.strip()
+        if stripped_tfdata[0] in ['{', '['] and stripped_tfdata[-1] in ['}', ']']:
+            filetype = 'json'
+            loaded_template_data = json.loads(tfdata, object_pairs_hook=collections.OrderedDict)
+        else:
+            filetype = 'yaml'
+            loaded_template_data = cfy.ordered_safe_load(open(filename, 'rU'), object_pairs_hook=collections.OrderedDict)
+        return (filetype, loaded_template_data, tfdata)
+
+    def write(self):
```
Comment (jaymccon) [inline taskcat/amiupdater.py]: why do we need this no-op method ?

File: taskcat/amiupdater.py [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+
+class TemplateObject(TemplateClass):
+    _objs = []
+    replacement_ami_map = {}
+
+    @classmethod
+    def objects(cls):
+        return cls._objs
+
+    def __init__(self, filename, all_regions=False):
```
Comment (jaymccon) [inline taskcat/amiupdater.py]: `__init__` method is pretty big, with several loops and conditions, may be more readable to split some of it out into private methods

File: taskcat/amiupdater.py (hunk 5 above)
Comment (andrew-glenn) [reply taskcat/amiupdater.py]: Whoops. that's a carry-over from development. Will remove it.

**Comment by andrew-glenn** [comment]
@jaymccon  - Assuming you approve, please *DO NOT* merge; I want to squash everything down before this hits develop.

File: taskcat/amiupdater.py [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+        for _ec2r in _ec2_regions:
+            self._region_list.append(_ec2r['RegionName'])
+        if self.all_regions:
+            for region in self._region_list:
+                self._regions.add(region)
+        else:
+            # Use the regions that are in Mappings/AWSAMIRegionMap
+            for region in self._mapping_root.keys():
+                if region not in self._region_list:
+                    raise AMIUpdaterException("Template: [{}] Region: [{}] is not a valid region".format(self._filename, region))
```
Comment (jaymccon) [inline taskcat/amiupdater.py]: this breaks if the legacy `AMI` key is present

amiupdater ~/quickstart-amazon-eks/templates/amazon-eks-nodegroup.template.yaml
Traceback (most recent call last):
  File "/Users/jmmccon/taskcat/venv-pc/bin/amiupdater", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/Users/jmmccon/taskcat/bin/amiupdater", line 38, in <module>
    a.update_amis()
  File "/Users/jmmccon/taskcat/taskcat/amiupdater.py", line 390, in update_amis
    TemplateObject(template_file)
  File "/Users/jmmccon/taskcat/taskcat/amiupdater.py", line 252, in __init__
    self._determine_regions()
  File "/Users/jmmccon/taskcat/taskcat/amiupdater.py", line 320, in _determine_regions
    raise AMIUpdaterException("Template: [{}] Region: [{}] is not a valid region".format(self._filename, region))
taskcat.amiupdater.AMIUpdaterException: Template: [/Users/jmmccon/quickstart-amazon-eks/templates/amazon-eks-nodegroup.template.yaml] Region: [AMI] is not a valid region

File: taskcat/amiupdater.py [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+            else:
+                raw_ami_names.update({rcn.cn:
+                                          {rcn.region: [APIResultsData(rcn.cn, x['ImageId'],
+                                                                       int(datetime.datetime.strptime(x['CreationDate'],
+                                                                                                      "%Y-%m-%dT%H:%M:%S.%fZ").timestamp()), rcn.region)
+                                                        for x in rcn.results]}})
+
+        for codename, regions in raw_ami_names.items():
+            for region, results_list in regions.items():
+                latest_ami = sorted(results_list, reverse=True)[0]
```
Comment (jaymccon) [inline taskcat/amiupdater.py]: should check the length of the list first to determine if there are any results and avoid this exception:

```
amiupdater ~/quickstart-amazon-eks/templates/amazon-eks-nodegroup.template.yaml
[INFO   ] : Created all filters necessary for the API calls
[INFO   ] : Latest AMI IDs fetched
Traceback (most recent call last):
  File "/Users/jmmccon/taskcat/venv-pc/bin/amiupdater", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/Users/jmmccon/taskcat/bin/amiupdater", line 38, in <module>
    a.update_amis()
  File "/Users/jmmccon/taskcat/taskcat/amiupdater.py", line 398, in update_amis
    Codenames.parse_api_results()
  File "/Users/jmmccon/taskcat/taskcat/amiupdater.py", line 155, in parse_api_results
    latest_ami = sorted(results_list, reverse=True)[0]
IndexError: list index out of range
```

**Comment by jaymccon** [comment]
LGTM, great stuff @andrew-glenn !

**Pull request merged**